components/places/src/bookmark_sync/store.rs

```diff
@@ -951,100 +1372,98 @@ impl<'a> dogear::Store<Error> for Merger<'a> {
+        if ops.is_empty() {
+            // If we don't have any items to apply, upload, or delete,
+            // no need to open a transaction at all.
+            return Ok(());
         }
 
-        Ok(contents)
-    }
-
-    fn apply<'t>(
-        &mut self,
-        root: MergedRoot<'t>,
-        deletions: impl Iterator<Item = Deletion<'t>>,
-    ) -> Result<()> {
-        self.store.interruptee.err_if_interrupted()?;
-        let descendants = root.descendants();
-
-        self.store.interruptee.err_if_interrupted()?;
-        let deletions = deletions.collect::<Vec<_>>();
-
         let tx = if !self.external_transaction {
             Some(self.store.db.begin_transaction()?)
         } else {
             None
         };
+
+        log::debug!("Updating local items in Places");
         self.store
-            .update_local_items(self.local_time, descendants, deletions)?;
-        self.store.stage_local_items_to_upload()?;
-        self.store.db.execute_batch(
-            "DELETE FROM mergedTree;
-             DELETE FROM idsToWeaklyUpload;",
-        )?;
+            .update_local_items_in_places(self.local_time, &ops)?;
+
+        log::debug!("Staging items to upload");
+        self.store
+            .stage_items_to_upload(&ops.upload_items, &ops.upload_tombstones)?;
+
+        self.store.db.execute_batch("DELETE FROM itemsToApply;")?;
         if let Some(tx) = tx {
             tx.commit()?;
         }
         Ok(())
     }
 }
 
+/// Formats an optional value so that it can be included in a SQL statement.
+struct NullableFragment<T>(Option<T>);
+
+impl<T> fmt::Display for NullableFragment<T>
+where
+    T: fmt::Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.0 {
+            Some(v) => v.fmt(f),
+            None => write!(f, "NULL"),
+        }
+    }
+}
+
+/// Formats a `SELECT` statement for staging local items in the `itemsToUpload`
+/// table.
+struct UploadItemsFragment(&'static str, &'static str);
+
+impl fmt::Display for UploadItemsFragment {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "SELECT {alias}.id, {alias}.guid, {alias}.syncChangeCounter,
+                    p.guid AS parentGuid, p.title AS parentTitle,
+                    {alias}.dateAdded / 1000 AS dateAdded,
+                    (CASE b.type
+                     WHEN {bookmark_typ} THEN
+                         CASE substr(h.url, 1, 6)
+                         WHEN 'place:' THEN {query_kind}
+                         ELSE {bookmark_kind}
+                         END
+                     WHEN {folder_typ} THEN {folder_kind}
+                     WHEN {separator_typ} THEN {separator_kind}
+                     END),
+                    {alias}.title,
+                    h.id AS placeId,
+                    h.url,
+                    (SELECT v.keyword FROM moz_bookmarks_synced v
```

Comment:
This is kind of icky; we fetch the keyword out of the mirror table, and stuff it right back into the record for uploading. But, if we decide to store keywords in Places, too (#2472), we can remove this.